app/flatpak-cli-transaction.c

```diff
@@ -149,196 +163,310 @@ op_type_to_string (FlatpakTransactionOperationType operation_type)
     }
 }
 
-#define BAR_LENGTH 20
-#define BAR_CHARS " -=#"
-
 static void
-progress_done (FlatpakTransaction *transaction)
+redraw (FlatpakCliTransaction *self)
 {
-  FlatpakCliTransaction *self = FLATPAK_CLI_TRANSACTION (transaction);
+  int top;
+  int row;
+  int col;
+  int skip;
 
-  if (self->progress_initialized)
+  top = self->end_row - self->table_height;
+  if (top > 0)
     {
-      g_print ("\n");
-      self->progress_initialized = FALSE;
+      row = top;
+      skip = 0;
+    }
+  else
+    {
+      row = 1;
+      skip = 1 - top;
     }
+
+  g_print (FLATPAK_ANSI_ROW_N FLATPAK_ANSI_CLEAR, row);
+  // we update table_height and end_row here, since we might have added to the table
+  flatpak_table_printer_print_full (self->printer, skip, self->cols,
+                                    &self->table_height, &self->table_width);
+  flatpak_get_cursor_pos (&self->end_row, &col);
+  self->end_row += 1;
+}
+
+static void
+set_op_progress (FlatpakCliTransaction *self,
+                 FlatpakTransactionOperation *op,
+                 char progress)
+{
+  if (flatpak_fancy_output ())
+    {
+      int row = GPOINTER_TO_INT (g_object_get_data (G_OBJECT (op), "row"));
+      char cell[] = "[ ]";
+      cell[1] = progress;
+      flatpak_table_printer_set_cell (self->printer, row, 1, cell);
+    }
+}
+
+static void
+spin_op_progress (FlatpakCliTransaction *self,
+                  FlatpakTransactionOperation *op)
+{
+  const char p[] = "/-\\|/-\\|";
```

Comment:
Why two copies of the / - \ - | progression?

Reply:
oversight